### DISCUSSION TITLE
Kotlin

### DISCUSSION THREAD

**Discussion by kobylynskyi** [kobylynskyi/graphql-java-codegen#470] [General]
![image](https://user-images.githubusercontent.com/2664706/103145182-5c0ca400-46fb-11eb-87d7-eabca061a967.png)

Everything about Kotlin can be discussed here.

**Comment by kobylynskyi** [comment]
@jxnu-liguobin is Kotlin codegen ready to be shipped? Can I release it in [4.1.0](https://github.com/kobylynskyi/graphql-java-codegen/milestone/27?closed=1)?

**Reply by kobylynskyi** [reply]
The first version of graphql-java-codegen library with Kotlin support was released today:
### [4.1.0](https://github.com/kobylynskyi/graphql-java-codegen/releases/tag/v4.1.0)

**Reply by jxnu-liguobin** [reply]
Hi, @kobylynskyi ,
When I wrote the client, I found that when `apiInterfaceStrategy` and `apiRootInterfaceStrategy` are both set to `DO_NOT_GENERATE`, `apiNamePrefix` must be set, but it seems that `apiNamePrefix` is not used?

schema:
```
type Query {
  question(titleSlug: String!): QuestionNode!
}
```
I will get a Query resolver
```kotlin
interface QueryResolver {

    fun question(titleSlug: String): QuestionNodeTO

}
```

It seems that the setting of `apiNamePrefix` is redundant at this point

**Comment by chemeugene** [comment]
Hello!

Maybe it needs to have ability add @JvmStatic annotation on builder() method?
I have troubles with my attempts to do mapping of GraphQL generated dto to model objects and vice verca with Mapstruct 1.4.2

I use following task for generation dto:
```
task graphqlCodegenKotlinService(type: GraphQLCodegenGradleTask) {
    outputDir = new File("$buildDir/generatedGraphQL")
    generateClient = true
    generateApis = true
    generatedLanguage = GeneratedLanguage.KOTLIN
    generateBuilder = true
    generateImmutableModels = false
    generateModelOpenClasses = true
    generateDataFetchingEnvironmentArgumentInApis = true
    parentInterfaces {
        queryResolver = "graphql.kickstart.tools.GraphQLQueryResolver"
        mutationResolver = "graphql.kickstart.tools.GraphQLMutationResolver"
        subscriptionResolver = "graphql.kickstart.tools.GraphQLSubscriptionResolver"
        resolver = "graphql.kickstart.tools.GraphQLResolver<{{TYPE}}>"
    }
    customTypesMapping = [
            DateTime: "java.time.ZonedDateTime",
            JSON: "java.util.Map<java.lang.String,java.lang.String>",
            BigDecimal: "java.math.BigDecimal"
    ]
    customAnnotationsMapping = [
            "Character": ["@com.fasterxml.jackson.annotation.JsonTypeInfo(use=com.fasterxml.jackson.annotation.JsonTypeInfo.Id.NAME, include=com.fasterxml.jackson.annotation.JsonTypeInfo.As.PROPERTY,property = \"__typename\")",
                          "@com.fasterxml.jackson.annotation.JsonSubTypes(value = arrayOf(" + System.lineSeparator() +
                                  "       com.fasterxml.jackson.annotation.JsonSubTypes.Type(value = HumanTO::class, name = \"Human\"), " + System.lineSeparator() +
                                  "       com.fasterxml.jackson.annotation.JsonSubTypes.Type(value = DroidTO::class, name = \"Droid\")))"],
    ]
}
```

Also may be make sense to generated zero argument constructor.

**Reply by jxnu-liguobin** [reply]
Hi， @chemeugene Thank you for your report.
Do you need to use it in Java? That really needs to add this annotation.

Mapstruct is Java library...

Because adding this method will make it more tightly coupled with Java, at present, we try to add this kind of thing as little as possible.But if you really need it (if a JavaBean like library cannot be found in kotlin), you can create a PR and use the switch to control it.

@kobylynskyi What about you?
